language-server/src/dotty/tools/languageserver/Worksheet.scala

```diff
@@ -0,0 +1,96 @@
+package dotty.tools.languageserver
+
+import dotty.tools.dotc.ast.tpd.{DefTree, Template, Tree, TypeDef}
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.interactive.SourceTree
+import dotty.tools.dotc.util.Positions.Position
+import dotty.tools.dotc.util.SourceFile
+
+import dotty.tools.dotc.core.Flags.Synthetic
+
+import dotty.tools.repl.{ReplDriver, State}
+
+import org.eclipse.lsp4j.jsonrpc.CancelChecker
+
+import java.io.{ByteArrayOutputStream, PrintStream}
+import java.util.concurrent.{Callable, CancellationException, Executors, TimeUnit}
+
+object Worksheet {
+
+  private val executor = Executors.newFixedThreadPool(1)
+  private final val checkCancelledDelayMs = 50
+
+  /**
+   * Evaluate `tree` as a worksheet using the REPL.
+   *
+   * @param tree          The top level object wrapping the worksheet.
+   * @param sendMessage   A mean of communicating the results of evaluation back.
+   * @param cancelChecker A token to check whether execution should be cancelled.
+   */
+  def evaluate(tree: SourceTree,
+               sendMessage: String => Unit,
+               cancelChecker: CancelChecker)(
+      implicit ctx: Context): Unit = {
+    tree.tree match {
+      case td @ TypeDef(_, template: Template) =>
+        val replOut = new ByteArrayOutputStream
+        val repl = new ReplDriver(replOptions, out = new PrintStream(replOut))
+        val executed = collection.mutable.Set.empty[(Int, Int)]
+
+        template.body.foldLeft(repl.initialState) {
+          case (state, statement: DefTree) if statement.symbol.is(Synthetic) =>
+            state
+
+          case (state, statement) if executed.add(bounds(statement.pos)) =>
+            val task = executor.submit(new Callable[State] {
+              override def call(): State = {
+                  val (line, newState) = execute(repl, state, statement, tree.source)
+                  val result = new String(replOut.toByteArray())
+                  if (result.trim.nonEmpty) sendMessage(encode(result, line))
+                  replOut.reset()
+                  newState
+              }
+            })
+
+            while (!task.isDone()) {
+              try {
+                cancelChecker.checkCanceled()
+                Thread.sleep(checkCancelledDelayMs)
+              } catch {
+                case _: CancellationException =>
+                  task.cancel(true)
+              }
+            }
+
+            try task.get(0, TimeUnit.SECONDS)
```

Comment:
Maybe make this configurable, the default can be infinite. Otherwise bound it to the configured timeout.